ports/raspberrypi/common-hal/busio/UART.c

```diff
@@ -24,238 +24,175 @@
+
+    uint8_t uart_id = get_free_uart();
+
+    self->tx_pin = pin_init(uart_id, tx, 0);
+    self->rx_pin = pin_init(uart_id, rx, 1);
+    self->cts_pin = pin_init(uart_id, cts, 2);
+    self->rts_pin = pin_init(uart_id, rts, 3);
+
+    self->uart = UART_INST(uart_id);
+    self->uart_id = uart_id;
+    self->baudrate = baudrate;
+    self->timeout_ms = timeout * 1000;
+
+    uart_init(self->uart, self->baudrate);
+    uart_set_fifo_enabled(self->uart, false);
+    uart_set_format(self->uart, bits, stop, parity);
+    uart_set_hw_flow(self->uart, (cts != NULL), (rts != NULL));
+
+    if (rx != NULL) {
+        // Initially allocate the UART's buffer in the long-lived part of the
+        // heap. UARTs are generally long-lived objects, but the "make long-
+        // lived" machinery is incapable of moving internal pointers like
+        // self->buffer, so do it manually.  (However, as long as internal
+        // pointers like this are NOT moved, allocating the buffer
+        // in the long-lived pool is not strictly necessary)
+        // (This is a macro.)
+        if (!ringbuf_alloc(&ringbuf[uart_id], receiver_buffer_size, true)) {
+            mp_raise_msg(&mp_type_MemoryError, translate("Failed to allocate RX buffer"));
+        }
+        if (uart_id) {
+            self->uart_irq_id = UART1_IRQ;
+            irq_set_exclusive_handler(self->uart_irq_id, uart1_callback);
+        } else {
+            self->uart_irq_id = UART0_IRQ;
+            irq_set_exclusive_handler(self->uart_irq_id, uart0_callback);
+        }
+        irq_set_enabled(self->uart_irq_id, true);
+        uart_set_irq_enables(self->uart, true, false);
+    }
 }
 
 bool common_hal_busio_uart_deinited(busio_uart_obj_t *self) {
-    return self->rx_pin == NO_PIN && self->tx_pin == NO_PIN;
+    return self->tx_pin == NO_PIN && self->rx_pin == NO_PIN;
 }
 
 void common_hal_busio_uart_deinit(busio_uart_obj_t *self) {
     if (common_hal_busio_uart_deinited(self)) {
         return;
     }
-    // This assignment is only here because the usart_async routines take a *const argument.
-    // struct usart_async_descriptor * const usart_desc_p = (struct usart_async_descriptor * const) &self->usart_desc;
-    // usart_async_disable(usart_desc_p);
-    // usart_async_deinit(usart_desc_p);
-    reset_pin_number(self->rx_pin);
+    uart_deinit(self->uart);
+    ringbuf_free(&ringbuf[self->uart_id]);
     reset_pin_number(self->tx_pin);
-    self->rx_pin = NO_PIN;
+    reset_pin_number(self->rx_pin);
+    reset_pin_number(self->cts_pin);
+    reset_pin_number(self->rts_pin);
     self->tx_pin = NO_PIN;
+    self->rx_pin = NO_PIN;
+    self->cts_pin = NO_PIN;
+    self->rts_pin = NO_PIN;
+}
+
+// Write characters.
+size_t common_hal_busio_uart_write(busio_uart_obj_t *self, const uint8_t *data, size_t len, int *errcode) {
+    if (self->tx_pin == NO_PIN) {
+        mp_raise_ValueError(translate("No TX pin"));
+    }
+
+    while (len > 0) {
+        if (uart_is_writable(self->uart)) {
```

Comment:
With the FIFO enabled you may be able to write up to 12 bytes immediately. Do this will allow more to be transmitted while background tasks run.

```suggestion
        while (uart_is_writable(self->uart) && len > 0) {
```